### PR TITLE
Object detection

### PR DESCRIPTION
**Pull request by kperrynrel**
Working PR for merging the mounting config object detection model into master.

### CODE REVIEW

File: panel_segmentation/panel_detection.py [hunk 1]
```diff
@@ -10,529 +12,677 @@
...
+    def classifyMountingConfiguration(self, image_file_path,
+                                      acc_cutoff=.65,
+                                      file_name_save=None,
+                                      use_nms=True):
+        """
+        This function is used to detect and classify the mounting configuration
+        of solar installations in satellite imagery. It leverages the Detecto
+        package's functionality
+        (https://detecto.readthedocs.io/en/latest/api/index.html),
+        to perform object detection on a satelllite image.
```
Comment (kanderso-nrel) [inline panel_segmentation/panel_detection.py:127]:
```suggestion
        to perform object detection on a satellite image.
```

File: panel_segmentation/panel_detection.py [hunk 2]
```diff
@@ -10,529 +12,677 @@
...
+
         Parameters
         -----------
-        latitude: Float. Latitude coordinate of the site.
-        longitude: Float. Longitude coordinate of the site.
-        file_name_save: String. File path that we want to save the image to. PNG file.
-        google_maps_api_key: String. Google Maps API Key for automatically 
-            pulling satellite images.
-                
+        latitude: (float)
```
Comment (kanderso-nrel) [inline panel_segmentation/panel_detection.py:67]: The parentheses around parameter types are visual clutter IMHO.  Note that sphinx adds parentheses automatically:

image

File: panel_segmentation/panel_detection.py [hunk 3]
```diff
@@ -541,74 +691,224 @@ def clusterPanels(self, test_mask, fig=False):
...
+        az_list_updated = list()
+        mounting_config_list_updated = list()
+        for array_number in range(len(az_list)):
+            mount_classification = mounting_config_list[array_number]
+            az = az_list[array_number]
+            if mount_classification == 'ground-single_axis_tracker':
+                if az <= 120:
+                    az = az + 90
+                elif az > 200:
+                    az = az - 90
```
Comment (kanderso-nrel) [inline panel_segmentation/panel_detection.py:903]: Maybe not a big deal, but our friends in the southern hemisphere may prefer an axis azimuth of zero over 180

File: panel_segmentation/panel_train.py [hunk 4]
```diff
@@ -13,372 +13,501 @@
...
+                                       NO_OF_TRAINING_IMAGES//self.BATCH_SIZE),
+                                   validation_data=val_image_generator,
+                                   validation_steps=(
+                                       NO_OF_VAL_IMAGES//self.BATCH_SIZE),
+                                   callbacks=[checkpoint]
                                    )
         return custom_model, results
-    
 
     def trainPanelClassifier(self, TRAIN_PATH, VAL_PATH,
```
Comment (kanderso-nrel) [inline panel_segmentation/panel_train.py:235]: PEP8 wouldn't like these all-caps parameter names

File: panel_segmentation/panel_train.py [hunk 5]
```diff
@@ -13,372 +13,501 @@
...
+                        ...images/
+                            ......a_image_1.png
+                            ......a_image_2.png
+                        ...annotations/
+                            ......b_image_1.xml
+                            ......b_image_2.xml
+        device: (string)
+            This argument is passed to the Model() class in Detecto.
+            It determines how to run the model: either on GPU via Cuda
+            (default setting), or on CPU. Please not that running the
```
Comment (kanderso-nrel) [inline panel_segmentation/panel_train.py:357]:
```suggestion
            (default setting), or on CPU. Please note that running the
```

File: setup.py [hunk 6]
```diff
@@ -36,9 +36,13 @@
 
 EXTRAS_REQUIRE = {
     'doc': [
-        'sphinx==1.8.5',
-        'sphinx_rtd_theme>=0.5.0',
+        'sphinx==3.2',
+        'jinja2<3.1',
+        'nbsphinx==0.8.8',
+        'nbsphinx-link==1.3.0',
```
Comment (kanderso-nrel) [inline setup.py:42]: Are the two nbsphinx packages actually needed here?  Doesn't seem like the docs use any notebooks.

**Review by kanderso-nrel — COMMENTED** [review]
Here are some notes; the actual package code is all over my head so I just skimmed that and flagged the few things that happened to jump out at me.  

One other thing: this PR currently has a couple .pyc files that should probably get dropped.

File: panel_segmentation/panel_train.py (hunk 4 above)
Comment (kperrynrel) [reply panel_segmentation/panel_train.py]: Removed all-caps variables

File: panel_segmentation/tests/test_panel_segmentation.py [hunk 7]
```diff
@@ -102,24 +110,27 @@ def test_crop_panels():
...
     new_res = pc.cropPanels(x, res)
-    #Assert that the 'new_res' variable is a numpy array and the dimensions.
+    # Assert that the 'new_res' variable is a numpy array and the dimensions.
     assert (type(new_res) == np.ndarray) & (new_res.shape == (1, 640, 640, 3))
 
-    
+
 def test_estimate_az():
     """
     Test the detectAzimuth() function.
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/tests/test_panel_segmentation.py]: Removed commenting + updated naming conventions

File: setup.py (hunk 6 above)
Comment (kperrynrel) [reply setup.py]: Addressed in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/ddf30adf67b87182aedaf527ba4febc0b1cbab9f, I removed these guys (they were initially copied over from rdtools

File: panel_segmentation/panel_detection.py (hunk 2 above)
Comment (kperrynrel) [reply panel_segmentation/panel_detection.py]: Removed these parentheses in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/c11b781b929615fa3d5646f42266f0285f8b0547

File: setup.py [hunk 8]
```diff
@@ -17,17 +17,17 @@
...
-	'scikit_image>=0.16.2',
-	'matplotlib>=3.1.3',
-	'requests>=2.22.0',
-	'tensorflow>=2.2.0',
-	'pandas>=1.1.4',
-	'cx_Freeze>=6.4.2',
-	'Pillow>=8.0.1',
-	'scikit_learn>=0.23.2',
-	'h5py>=2.10.0'
+    'opencv-python-headless==4.5.4.60',
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply setup.py:20]: This one was specifically erroring out on newer versions, so this was done on purpose

File: panel_segmentation/_version.py [hunk 9]
```diff
@@ -23,9 +23,9 @@ def get_keywords():
     # setup.py/versioneer.py will grep for the variable names, so they must
     # each be defined on a line of their own. _version.py will just call
     # get_keywords().
-    git_refnames = "$Format:%d$"
-    git_full = "$Format:%H$"
-    git_date = "$Format:%ci$"
+    git_refnames = " (HEAD -> master, object_detection)"
+    git_full = "98aef0a2a7abe9e7fc5b1003ba98752f3f19b721"
+    git_date = "2021-05-25 17:39:10 -0600"
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/_version.py]: Removed this by commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/c11b781b929615fa3d5646f42266f0285f8b0547, not sure why it was hardcoded here

File: setup.py [hunk 10]
```diff
@@ -17,17 +17,17 @@
...
+    'numpy>=1.18.1',
+    'scikit_image>=0.16.2',
+    'matplotlib>=3.1.3',
+    'requests>=2.22.0',
+    'tensorflow>=2.2.0',
+    'pandas>=1.1.4',
+    'Pillow>=8.0.1',
+    'scikit_learn>=0.23.2',
+    'h5py>=2.10.0',
+    'detecto>=1.2.1'
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply setup.py:30]: Added in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/c11b781b929615fa3d5646f42266f0285f8b0547, this was an error

File: panel_segmentation/tests/test_panel_segmentation.py [hunk 11]
```diff
@@ -129,25 +140,57 @@ def test_estimate_az():
...
-    res = pc.testSingle(x.astype(float), test_mask=None,  model =None)
-    #Crop the panels 
+    # Mask the satellite image
+    res = pc.testSingle(x.astype(float), test_mask=None,  model=None)
+    # Crop the panels
     new_res = pc.cropPanels(x, res)
     az = pc.detectAzimuth(new_res)
-    #Assert that the azimut returned is an int instance
+    # Assert that the azimut returned is an int instance
     assert_isinstance(az, float)
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/tests/test_panel_segmentation.py:102]: Addressed in https://github.com/NREL/Panel-Segmentation/pull/15/commits/8a77fed9b9a60c422c18e32bc9a1ab4d37281e62

File: panel_segmentation/tests/test_panel_segmentation.py [hunk 12]
```diff
@@ -188,23 +234,57 @@ def test_cluster_panels():
...
+        file_name_save_mount=None,
+        file_path_save_azimuth=None,
+        generate_image=False)
+    # Assert that a dictionary is returned with specific
+    # attributes
+    assert (type(site_analysis_dict) == dict) & \
+        (all([label == 'carport-fixed' for label in
+              site_analysis_dict["mounting_type"]])) & \
+        (sorted(site_analysis_dict['associated_azimuths']) ==
+         [90.0, 91.0, 161.0, 179.0])
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/tests/test_panel_segmentation.py]: Split all of the assert statements in https://github.com/NREL/Panel-Segmentation/pull/15/commits/8a77fed9b9a60c422c18e32bc9a1ab4d37281e62

File: panel_segmentation/tests/test_panel_train.py [hunk 13]
```diff
@@ -43,24 +39,28 @@ def test_train_panel_classifier():
...
     os.remove('./panel_segmentation/tests/classifier.h5')
-    #Assert the mod and results types.
-    assert (str(type(mod)) == "<class 'tensorflow.python.keras.engine.training.Model'>") & \
-            (str(type(results)) == "<class 'tensorflow.python.keras.callbacks.History'>")
-    
+    # Assert the mod and results types.
+    assert (str(type(mod)) ==
+            "<class 'tensorflow.python.keras.engine.training.Model'>") & \
+        (str(type(results)) ==
+         "<class 'tensorflow.python.keras.callbacks.History'>")
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/tests/test_panel_train.py]: Updated to assert_isinstance in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/8a77fed9b9a60c422c18e32bc9a1ab4d37281e62

File: sphinx/source/api.rst [hunk 14]
```diff
@@ -23,15 +23,17 @@ These classes perform analyses used in the Panel-Segmentation project.
    panel_detection.PanelDetection.cropPanels
    panel_detection.PanelDetection.plotEdgeAz
    panel_detection.PanelDetection.clusterPanels
-   
+   panel_detection.PanelDetection.runSiteAnalysisPipeline 
+   panel_detection.PanelDetection.classifyMountingConfiguration
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply sphinx/source/api.rst]: Added this in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/156943f7a5b6cdda2beb49e183f128e85e496104

File: setup.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kperrynrel) [reply setup.py]: Yep, added in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/8660e7cca853acd45763052982b5674b5cb38fa7

File: sphinx/source/api.rst [hunk 15]
```diff
@@ -43,3 +45,4 @@ The following deep learning models are included in the Panel-Segmentation packag
 
    VGG16_classification_model.h5
    VGG16Net_ConvTranpose_complete.h5
+   object_detection_model.pth
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply sphinx/source/api.rst:49]: @kanderso-nrel should I just remove the entire 'Models' section here?

File: panel_segmentation/panel_detection.py [hunk 16]
```diff
@@ -10,529 +12,677 @@
...
-        # r.content gives content, 
-        # in this case gives image 
-        f.write(r.content)   
-        # close method of file object 
-        # save and close the file 
+            raise ValueError("Response status code " +
+                             str(r.status_code) +
+                             ": Image not pulled successfully from API.")
+        # wb mode is stand for write binary mode
+        f = open(file_name_save, 'wb')
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/panel_detection.py]: Updated in https://github.com/NREL/Panel-Segmentation/pull/15/commits/97e1d2a8b9155cf86995c4e799513be3bfc695c2

File: panel_segmentation/tests/test_panel_segmentation.py [hunk 17]
```diff
@@ -25,22 +21,28 @@ def test_generate_satellite_image():
     -------
     None.
 
-    """    
+    """
     with pytest.raises(ValueError):
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/tests/test_panel_segmentation.py]: Updated this in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/b7009d15e10450eef683272d89ca6ae8ef06734b

File: panel_segmentation/tests/test_panel_segmentation.py [hunk 18]
```diff
@@ -52,20 +54,23 @@ def test_has_panels():
...
-                        color_mode='rgb', 
-                        target_size=(640,640))
+    # Pick the file name to read
+    img_file = \
+        "./panel_segmentation/examples/Panel_Detection_Examples/sat_img.png"
+    # Create an instance of the PanelDetection() class.
+    pc = pan_det.PanelDetection(
+        './panel_segmentation/VGG16Net_ConvTranpose_complete.h5',
+        './panel_segmentation/VGG16_classification_model.h5',
+        './panel_segmentation/object_detection_model.pth')
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/tests/test_panel_segmentation.py]: Converted to fixtures in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/9e05dceceab28c52c9a298a270216869a3d670f5

File: panel_segmentation/tests/test_panel_segmentation.py [hunk 19]
```diff
@@ -52,20 +54,23 @@ def test_has_panels():
...
+    # Read in the image
+    x = imagex.load_img(img_file,
+                        color_mode='rgb',
+                        target_size=(640, 640))
     x = np.array(x)
-    #Assert that the returned value is a boolean
+    # Assert that the returned value is a boolean
     panel_loc = pc.hasPanels(x)
     assert_isinstance(panel_loc, bool)
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/tests/test_panel_segmentation.py]: Tested specific value as true in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/af7be36715342b8cc731a2bd3112f0247c1e5b40

File: panel_segmentation/tests/test_panel_segmentation.py [hunk 20]
```diff
@@ -188,23 +234,57 @@ def test_cluster_panels():
...
     new_res = pc.cropPanels(x, res)
     n, clusters = pc.clusterPanels(new_res)
     azimuth_list = []
     for ii in np.arange(clusters.shape[0]):
-        az = pc.detectAzimuth(clusters[ii][np.newaxis,:])
+        az = pc.detectAzimuth(clusters[ii][np.newaxis, :])
         azimuth_list.append(az)
-    assert all(isinstance(float(x), float) for x in azimuth_list)
\ No newline at end of file
+    assert all(isinstance(float(x), float) for x in azimuth_list)
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply panel_segmentation/tests/test_panel_segmentation.py]: Changed this to testing specific values in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/af7be36715342b8cc731a2bd3112f0247c1e5b40

File: sphinx/source/changelog.rst [hunk 21]
```diff
@@ -3,6 +3,28 @@
...
+Documentation
+~~~~~~~~~~~~~
+- Updated Sphinx Documentation to account for new functions.
+- Updated the Jupyter notebooks to reflect pipeline changes: adding in the mounting configuration detection classifier and running it on satellite imagery.
+
+Scripts
+~~~~~~~~~~~~~
+- Add the function trainMountingConfigClassifier() to the TrainPanelSegmentationModel() class.
+- Add the functions runSiteAnalysisPipeline() and classifyMountingConfiguration() to the PanelDetection() class.
+- Add unit testing for the trainMountingConfigClassifier(), runSiteAnalysisPipeline(), and trainMountingConfigClassifier() functions.
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply sphinx/source/changelog.rst]: Added this in in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/435db04ec2ed0dfcdf803f5bc3f1548bc949cd23

File: setup.py [hunk 22]
```diff
@@ -36,9 +38,11 @@
 
 EXTRAS_REQUIRE = {
     'doc': [
-        'sphinx==1.8.5',
-        'sphinx_rtd_theme>=0.5.0',
+        'sphinx==3.2',
+        'jinja2<3.1',
+        'sphinx_rtd_theme==0.5.2',
         'ipython',
+        'sphinx-gallery==0.8.1'
```
Comment (kanderso-nrel) [inline setup.py:45]: `sphinx-gallery` doesn't seem necessary (yet?)

File: sphinx/source/api.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (kanderso-nrel) [reply sphinx/source/api.rst:49]: If there's no docs for these models, yeah I'd say no point in having an `autosummary` section for them.  If they need to be documented (up to you), just describing them here in the RST might make sense.

File: sphinx/source/changelog.rst [hunk 23]
```diff
@@ -3,6 +3,29 @@
 
 Change Log
 ==========
+Version 0.0.2 (April 29, 2022)
```
Comment (kanderso-nrel) [inline sphinx/source/changelog.rst:6]: Update for whenever this PR gets merged and the new version tagged

File: sphinx/source/changelog.rst [hunk 24]
```diff
@@ -3,6 +3,29 @@
...
+~~~~~~~~~~~~~
+- Updated Sphinx Documentation to account for new functions.
+- Updated the Jupyter notebooks to reflect pipeline changes: adding in the mounting configuration detection classifier and running it on satellite imagery.
+
+Scripts
+~~~~~~~~~~~~~
+- Add the function :py:func:`panel_segmentation.panel_train.trainMountingConfigClassifier`.
+- Add the functions :py:func:`panel_segmentation.panel_detection.runSiteAnalysisPipeline` and :py:func:`panel_segmentation.panel_detection.classifyMountingConfiguration`.
+- Add unit testing for the :py:func:`panel_segmentation.panel_train.trainMountingConfigClassifier`, :py:func:`panel_segmentation.panel_detection.runSiteAnalysisPipeline`, 
+  and :py:func:`panel_segmentation.panel_detection.classifyMountingConfiguration` functions.
```
Comment (kanderso-nrel) [inline sphinx/source/changelog.rst:23]: None of these link correctly, I assume because they're missing the corresponding class names?

**Review by kanderso-nrel — COMMENTED** [review]
A bit more nitpicking :)

File: sphinx/source/changelog.rst (hunk 24 above)
Comment (kperrynrel) [reply sphinx/source/changelog.rst]: Yep--addressed in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/41ecea125c343fc3db41b057994ba949b468f77b

File: setup.py (hunk 22 above)
Comment (kperrynrel) [reply setup.py]: Removed dependency in commit https://github.com/NREL/Panel-Segmentation/pull/15/commits/41ecea125c343fc3db41b057994ba949b468f77b

**Pull request merged**